ITE: drivers/usb/device: Add USB Device Controller Support

bjchentc opened the pull request:
Add USB Device Driver (usb_dc) of ITE IT82xx2

TEST=west build -p always -b it82xx2_evb 
with the sample zephyr/sample/subsys/usb/hid

Signed-off-by: BJ Chen <bj.chen@ite.com.tw>

jfischer-no commented:
Please note, you implement a driver using the API we like to scrap ASAP.

itebj commented on include/zephyr/drivers/usb/it82xx2_usb_dc.h:
```diff
@@ -0,0 +1,192 @@
+/*
```
Thanks for reminding, sir.
I've removed the private header and move definitions to the C file.

Thanks again.

itebj replied:
@jfischer-no 

After moving related definitions to usb_dc_it82xx2.c, the private header file has been removed.
Thanks again for your guidance.

jfischer-no commented:
> > Please note, you implement a driver using the API we like to scrap ASAP.
> 
> Can you point to the new USB device controller API? What is the timeline for switching over to the new API?

There is meta-issue for USB support, https://github.com/zephyrproject-rtos/zephyr/issues/42066

itebj commented:
Dear sirs, 

I upload again for rebasing the upstream/main with adding the following lines in boards/riscv/it82xx2_evb/it82xx2_evb.dts:

1) sw1-sw3 aliases:
```
		sw1 = &button1;
		sw2 = &button2;
		sw3 = &button3;
```
2) button1-3
```
		button1: button_1 {
			gpios = <&gpioc 4 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
			label = "Push button switch 1";
		};
		button2: button_2 {
			gpios = <&gpioe 3 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
			label = "Push button switch 2";
		};
		button3: button_3 {
			gpios = <&gpioe 4 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
			label = "Push button switch 3";
		};
```
Thanks again for your re-visiting.

itebj commented:
> > > Please note, you implement a driver using the API we like to scrap ASAP.
> > 
> > 
> > Can you point to the new USB device controller API? What is the timeline for switching over to the new API?
> 
> There is meta-issue for USB support, #42066

Truly appreciative of your notifications, sir. I believe that we'll move on to follow the USB device "next" afterwards. However, due to the requirements of some running projects, we'd still have to use the potentially deprecated old implementation temporarily for current situation.

We'll start to study related and required information about the new USB support ASAP.

Thanks again for your kindly guidance & best regards,
BJ

itebj commented:
> > > Please note, you implement a driver using the API we like to scrap ASAP.
> > 
> > 
> > Can you point to the new USB device controller API? What is the timeline for switching over to the new API?
> 
> There is meta-issue for USB support, #42066

Truly appreciative of your notifications, sir. I believe that we'll move on to follow the USB device "next" afterwards. However, due to the requirements of some running projects, we'd still have to use the potentially deprecated old implementation temporarily for current situation.

We'll start to study related and required information about the new USB support ASAP.

Thanks again for your kindly guidance & best regards,
BJ

itebj commented:
Dear sirs,

Following #57028,  I made some modifications as follows:

`	if (ep_idx >= MAX_NUM_ENDPOINTS) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}

	if ((ep_idx != 0) && (!in && ep_idx % FIFO_NUM != 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}

	if ((ep_idx != 0) && (in && ep_idx % FIFO_NUM == 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}
`
Please revisit if you're available.
Thanks again & Best regards,
BJ

itebj commented:
Dear sirs,

Following https://github.com/zephyrproject-rtos/zephyr/pull/57028, I made some modifications as follows:

`
if (ep_idx >= MAX_NUM_ENDPOINTS) {
LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
return -EINVAL;
}
`
`
if ((ep_idx != 0) && (!in && ep_idx % FIFO_NUM != 2)) {
	LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
	return -EINVAL;
}

if ((ep_idx != 0) && (in && ep_idx % FIFO_NUM == 2)) {
	LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
	return -EINVAL;
}
`
Please revisit if you're available.
Thanks again & Best regards,
BJ

itebj commented:
Dear sirs,

This update is to follow the #57028, we made some modifications towards the usb_dc_ep_check_cap().

`
	if (ep_idx >= MAX_NUM_ENDPOINTS) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}

	if ((ep_idx != 0) && (!in && ep_idx % FIFO_NUM != 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}

	if ((ep_idx != 0) && (in && ep_idx % FIFO_NUM == 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}
`

Truly grateful for your revisiting.

Regards,
BJ

itebj commented:
Dear sirs,

This update is to follow the #57028, we made some modifications towards the usb_dc_ep_check_cap() as follows:

`	if ((ep_idx != 0) && (!in && ep_idx % FIFO_NUM != 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}

	if ((ep_idx != 0) && (in && ep_idx % FIFO_NUM == 2)) {
		LOG_WRN("Invalid Endpoint Number 0x%02x", cfg->ep_addr);
		return -EINVAL;
	}`

Truly grateful for your revisiting.

Regards,
BJ

itebj commented:
Dear sirs,

This update is to follow the #57028, we made some modifications towards two conditions in the usb_dc_ep_check_cap() as follows:

Condition 1.
`((ep_idx != 0) && (!in && ep_idx % FIFO_NUM != 2))`

Condition 2.
`((ep_idx != 0) && (in && ep_idx % FIFO_NUM == 2))`

The conditions above is about the chip-level FIFO assignment.

Truly grateful for your revisiting & Best regards,
BJ

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+	.wuc_list = usb_wuc0
+};
+
+struct usb_device_ep_data {
```
`struct it82xx2_endpoint_data {`

itebj replied:
Got it, sir. I've renamed it with prefix.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+};
+
+struct usb_device_ep_data {
+	/* Endpoint IN callback function */
```
Please remove all trivial comments, apply it everywhere.

itebj replied:
Got it, sir. I've removed most trivial comments in this upload. Thanks for your suggestions.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+	uint16_t mps;
+};
+
+struct usb_device_data {
```
`struct usb_it82xx2_data {`

itebj replied:
Likewise, I've modified it to be struct usb_it82xx2_data. Thank you, sir.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+ */
+static const uint8_t epn_ext_ctrl_tbl[4] = {1, 0, 1, 0};
+
+static usb_dc_status_callback usb_status_cb;
```
Can be part of `struct usb_it82xx2_data {`

itebj replied:
Yes, indeed, I've merge it into the struct usb_it82xx2_data.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+
+static usb_dc_status_callback usb_status_cb;
+static struct usb_device_data udata0 = {
+	.attached = 0,
+	.st_state = 0,
+	.no_data_ctrl = false,
+	.suspended = false,
```
Redundant initialization

itebj replied:
Thank you, sir. The redundancy has been removed in this upload.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+/*
+ * Abbreviations of the USB Slave Control Registers
+ */
+struct usb_it82xx2_regs *const usb_regs =
```
`static struct usb_it82xx2_regs *const it82xx2_base =`

Can be part of `struct usb_it82xx2_config`, and use DT_INST_REG_ADDR(0)

```c
struct usb_it82xx2_config ucfg0 = {
	.base = (struct usb_it82xx2_regs *)DT_INST_REG_ADDR(0),
};
```

itebj replied:
Yes, sir. I've put it into a part of the struct usb_it82xx2_config which uses DT_INST_REG_ADDR(0) in the initialization. Thanks again for your suggestions.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+struct usb_it82xx2_regs *const usb_regs =
+	(struct usb_it82xx2_regs *)IT82XX2_USB_BASE;
+
+struct it82xx2_usb_ep_regs *ep_regs =
+	usb_regs->usb_ep_regs;
+
+struct it82xx2_usb_ep_fifo_regs *ff_regs =
+	usb_regs->fifo_regs;
+
+volatile uint8_t *epn0n1_ext_ctrl =
+	usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.EPN0N1_EXT_CTRL;
+
+volatile uint8_t *ep_fifo_ctrl =
+	usb_regs->fifo_regs[EP_EXT_REGS_BX].fifo_ctrl.EP_FIFO_CTRL;
+
+struct epn_ext_ctrl_regs *epn_ext_ctrl =
+	usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.EPN_EXT_CTRL;
```
Missing static, should be rather implemented as helper functions.

itebj replied:
Got it, sir. Thanks for reminding, now I've made them all static.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+ *
+ * WU90 (USB D+) Enable/Disable
+ */
+void usb_it8xxx2_enable_wu90_irq(const struct device *dev, bool enable)
```
Missing static, please check everywhere.

itebj replied:
All the functions which should be static have been modified accordingly. Thanks you for your reminding, sir.

jfischer-no commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+	}
+}
+
+/*
+ * Function it82xx2_wu90_isr():
+ *
+ * WU90 Interrupt Service Routine
+ */
```
Please remove these unnecessary comments.
I stop here for first round,

itebj replied:
Got it, sir. Most unnecessary comments have been removed. Truly appreciative of your great patience, suggestions & guidance.

Thanks again & Best regards,
BJ

itebj commented on boards/riscv/it82xx2_evb/it82xx2_evb.dts:
```diff
@@ -40,6 +44,27 @@
+			label = "Push button switch 3";
+		};
+	};
+
```
Got it, sir. I've removed them in commit/PR.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+
+#define LOG_LEVEL CONFIG_USB_DRIVER_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usb_dc_it82xx2);
```
Thanks for suggestions, sir. The modifications have been made.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+ *
+ * USB Slave Controller Registers Bits & Constants
+ *
+ */
```
Got it, I've modified all the slave keywords into device except those registers defined in the chip datasheet.

itebj replied:
Sorry for my forgetting to shrink it to one line, now it's updated in the latest upload. Thanks again for your revisiting, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+/* The bit definitions of the register EP TRANSTYPE STATUS:
+ * 0x42/0x46/0x4A/0x4E
+ */
+#define ENDPOINT_READY_STATUS		BIT(2)
```
Indeed, some macros and definitions are unused, I've removed unused macros. Thanks again for your advice, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+#define COMPLETED_TRANS			0xF0
+
+/* EP Definitions */
+#define EP_IN_MASK			BIT(7)
```
Got it, sir. Now I just use the USB_EP_DIR_MASK defined in the usb_ch9.h. Thanks for your kindly reminding.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+#define EP_INVALID_MASK			~(EP_IN_MASK | EP_VALID_MASK)
+
+/* Bit [1:0] represents the TRANSACTION_TYPE as follows: */
+enum TRANSACTION_TYPE {
```
Got it. I've modified them to be prefixed with it82xx2. Thanks again for guidance, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+	 * At this level there's no difference between L1 and L2
+	 * suspend states.  (L2 being original USB 1.1 suspend.)
+	 */
+};
```
Unused definitions are removed in this upload. Thanks again.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1687 @@
+};
+
+enum trans_type {
+	NONE_TOKEN,
```
My blunder, it was a test piece and it shouldn't be there, now it's removed. Thanks for reminding.

itebj commented:
Additional modifications in this commit to fix some potential errors:

1. Fix the EPN_EXT_CTRL2 ready bit definition for extended endpoints:
`#define READY_BITS			0x0F`

2. exception handling:

`
static void ep_in_out_config(uint8_t idx)
{
	uint8_t ep_trans = (epn_ext_ctrl[idx].EPN_EXT_CTRL2 >> 4) & READY_BITS;

	if (udata0.ep_data[ep_trans].ep_status == EP_CONFIG_IN) {
		if (ep_trans < 4) {
			if (!!(ep_regs[idx].EP_CTRL & EP_OUTDATA_SEQ)) {
				ep_regs[idx].EP_CTRL &= ~EP_OUTDATA_SEQ;
			} else {
				ep_regs[idx].EP_CTRL |= EP_OUTDATA_SEQ;
			}
		} else {
			usb_extend_ep_ctrl(ep_trans, EXT_EP_DATA_SEQ_INV);
		}

		if (udata0.ep_data[ep_trans].cb_in) {
			udata0.ep_data[ep_trans].cb_in(ep_trans | 0x80,
				USB_DC_EP_DATA_IN);
		}

	} else {
		if (udata0.ep_data[ep_trans].cb_out) {
			udata0.ep_data[ep_trans].cb_out(ep_trans,
				USB_DC_EP_DATA_OUT);
		}
	}
}
`

Add the condition `if (udata0.ep_data[ep_trans].cb_in) {` & `if (udata0.ep_data[ep_trans].cb_out) {
` to avoid exceptions.

Thanks again for your great patience and revisiting.

tmon-nordic commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1564 @@
+ * Abbreviations of the USB Device Control Registers
+ */
+
+static struct usb_it82xx2_regs *const usb_regs =
```
What's the point in making these actual variables? Shouldn't these be just macros?

itebj replied:
Appreciative of your feedback, sir. 

Due to the fact that the inconsistency of the register mapping/definitions in datasheet, the registers definitions will lead to very long statements and it's a little deep in structure which could violate coding style rules for both Zephyr / Chromium. (100 cols for Zephyr, 80 cols for CrOS)  Please refer to line 479 of the chip_chipregs.h which depicts the registers definitions.

Thanks again.

itebj replied:
Yes, sir. It has been dropped.

itebj replied:
Yes, sir. This one has been dropped in the latest update.
Thanks again.

tmon-nordic commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1564 @@
+	}
+
+	if ((ctrl >= EXT_EP_DIR_IN) && (ctrl < EXT_EP_READY)) {
+	/* From EXT_EP_DIR_IN to EXT_EP_DISABLE */
```
nitpick: comment not aligned

itebj replied:
Thanks for reminding, sir. I'll fix it in the next upload.

tmon-nordic commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1564 @@
+		epn_ext_ctrl_cfg1(reg_idx, mask, flag);
+
+	} else if ((ctrl >= EXT_EP_ISO_DISABLE) && (ctrl < EXT_EP_DIR_IN)) {
+	 /* From EXT_EP_ISO_DISABLE to EXT_EP_DATA_SEQ_0 */
```
nitpick: comment not aligned

itebj replied:
Thanks again for reminding, sir. I'll fix it in the next upload.

tmon-nordic commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1564 @@
+	} else if ((ctrl >= EXT_EP_ISO_DISABLE) && (ctrl < EXT_EP_DIR_IN)) {
+	 /* From EXT_EP_ISO_DISABLE to EXT_EP_DATA_SEQ_0 */
+		reg_idx = (ep_idx - 4) >> 1;
+		flag = (ep_idx % 2 == 0) ? (false) : (true);
```
Is this equivalent to `flag = (ep_idx & 1);`?

itebj replied:
Yes, you're absolutely right. I should thought of it.Thanks for guidance, sir. I'll fix it in the next upload.

itebj replied:
Thanks for hint, sir.
Now I modify it to be 
`flag = !(ep_idx & 1) ? (false) : (true);`

tmon-nordic replied:
Why make it so complex? `flag = !!(ep_idx & 1);` should work just fine.

itebj replied:
Sorry that I misunderstood the question mark as a part of the ternary operator, now it's fixed as follows:

`flag = !!(ep_idx & 1);`


Since the flag is equal to the inverse of original (ep_idx % 2 == 0), so I swap the expr1 and expr2 when I use the flag as the condition as follows:

`mask = flag ? (ext_ctrl_tbl[ctrl] << 4) : (ext_ctrl_tbl[ctrl]);`

Thanks again for your great patience and helps, sir.

itebj commented:
Dear sirs,

This upload is also modified for resolving the ep_status being set wrong in usb_dc_ep_configure() as follows:
`
	if (ep_idx < EP4) {
		(in) ? (ep_regs[ep_idx].EP_CTRL |= EP_DIRECTION | EP_OUTDATA_SEQ) :
			(ep_regs[ep_idx].EP_CTRL &= ~EP_DIRECTION);

	} else {

		(in) ? (usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_IN)) :
			(usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_OUT));

		if (in) {
			usb_extend_ep_ctrl(ep_idx, EXT_EP_DATA_SEQ_1);
		}

		LOG_DBG("ep_status %d", udata0.ep_data[ep_idx].ep_status);
	}

	(in) ? (udata0.ep_data[ep_idx].ep_status = EP_CONFIG_IN) :
		(udata0.ep_data[ep_idx].ep_status = EP_CONFIG_OUT);

	ep_fifo_ctrl[reg_idx] |= reg_val;
`
Thanks again for your revisiting.

itebj commented:
Dear sirs,

This upload is also modified for resolving the ep_status being set wrong in usb_dc_ep_configure() as follows:

if (ep_idx < EP4) {
	(in) ? (ep_regs[ep_idx].EP_CTRL |= EP_DIRECTION | EP_OUTDATA_SEQ) :
		(ep_regs[ep_idx].EP_CTRL &= ~EP_DIRECTION);
} else {
	(in) ? (usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_IN)) :
		(usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_OUT));

	if (in) {
		usb_extend_ep_ctrl(ep_idx, EXT_EP_DATA_SEQ_1);
	}
	LOG_DBG("ep_status %d", udata0.ep_data[ep_idx].ep_status);
}

(in) ? (udata0.ep_data[ep_idx].ep_status = EP_CONFIG_IN) :
	(udata0.ep_data[ep_idx].ep_status = EP_CONFIG_OUT);

ep_fifo_ctrl[reg_idx] |= reg_val;

Thanks again for your revisiting.

itebj commented:
Dear sirs, 

Thanks again for your revisiting & time sharing.
This update is about fixing the ep_status being set wrongly in the usb_dc_ep_configure() as the following figure:

```
	if (ep_idx < EP4) {
		(in) ? (ep_regs[ep_idx].EP_CTRL |= EP_DIRECTION | EP_OUTDATA_SEQ) :
			(ep_regs[ep_idx].EP_CTRL &= ~EP_DIRECTION);

	} else {

		(in) ? (usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_IN)) :
			(usb_extend_ep_ctrl(ep_idx, EXT_EP_DIR_OUT));

		if (in) {
			usb_extend_ep_ctrl(ep_idx, EXT_EP_DATA_SEQ_1);
		}

		LOG_DBG("ep_status %d", udata0.ep_data[ep_idx].ep_status);
	}

	(in) ? (udata0.ep_data[ep_idx].ep_status = EP_CONFIG_IN) :
		(udata0.ep_data[ep_idx].ep_status = EP_CONFIG_OUT);

	ep_fifo_ctrl[reg_idx] |= reg_val;

jfischer-no reviewed (changes requested):
A few more comments.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+};
+
+struct usb_it82xx2_config {
+	uintptr_t base_addr;
```
Corrected as follows:
`struct usb_it82xx2_config {
	const struct usb_it82xx2_regs *base;`

itebj replied:
Thank you, sir. It has been corrected as you mentioned.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+PINCTRL_DT_INST_DEFINE(0);
+
+static const struct usb_it82xx2_config ucfg0 = {
+	.base_addr = DT_INST_REG_ADDR(0),
```
It has been modified as follows:
`.base = (struct usb_it82xx2_regs *)DT_INST_REG_ADDR(0),`

Thanks again for guidnace.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+	(struct usb_it82xx2_regs *)ucfg0.base_addr;
+
+static struct it82xx2_usb_ep_regs *ep_regs =
+	usb_regs->usb_ep_regs;
```
These abbreviations have been dropped by adding function `static struct usb_it82xx2_regs *it82xx2_get_usb_regs(void)`.

However, the following line is required for the most usb_dc APIs do not have the *dev arg:
`const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(usb0));`

itebj replied:
Truly grateful for for your guidance, sir.

These abbreviations have been dropped by adding function `static struct usb_it82xx2_regs *it82xx2_get_usb_regs(void)`.

However, the following line is required for the most usb_dc APIs do not have the *dev arg:
`const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(usb0));`

Thanks again.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+					if (ret != 0) {
+						continue;
+					}
+				}
```
These functions have been modified to be `it82xx2_setup_done(), it82xx2_in_done(), it82xx2_out_done()` which are invoked in `it82xx2_usb_dc_trans_done().`

Thanks again for guidance.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+{
+	uint8_t status = usb_regs->SLAVE_INTERRUPT_STATUS &
+		usb_regs->SLAVE_INTERRUPT_MASK; /* mask non enable int */
+
```
Removed. Thanks for reminding, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+		return 0;
+	}
+	LOG_DBG("Detached");
+	irq_disable(IT8XXX2_USB_IRQ);
```
It has been corrected. Thanks again for reminding, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+}
+
+#ifdef CONFIG_USB_DEVICE_REMOTE_WAKEUP
+bool usb_check_suspend(void)
```
Got it. I've changed it to be `it82xx2_check_suspend()` which I mistook it as one of the Driver APIs. Thanks again for your correction, sir.

itebj commented on drivers/usb/device/usb_dc_it82xx2.c:
```diff
@@ -0,0 +1,1540 @@
+	return udata0.ep_data[ep_idx].mps;
+}
+
+#ifdef CONFIG_USB_DEVICE_REMOTE_WAKEUP
```
Got it, sir. Thanks again for correction.

jfischer-no reviewed (changes requested):
LGTM, just two minor comments, please add zephyr_udc0 nodelabel and usb_device to boards/riscv/it82xx2_evb/it82xx2_evb.yaml.

```diff
diff --git a/boards/riscv/it82xx2_evb/it82xx2_evb.dts b/boards/riscv/it82xx2_evb/it82xx2_evb.dts
index 310aef562a0..e0cbf2e67e5 100644
--- a/boards/riscv/it82xx2_evb/it82xx2_evb.dts
+++ b/boards/riscv/it82xx2_evb/it82xx2_evb.dts
@@ -197,7 +197,7 @@
        pinctrl-names = "default";
 };
 
-&usb0 {
+zephyr_udc0: &usb0 {
        status = "okay";
        pinctrl-0 = <&usb0_dm_gph5_default
                     &usb0_dp_gph6_default>;
diff --git a/boards/riscv/it82xx2_evb/it82xx2_evb.yaml b/boards/riscv/it82xx2_evb/it82xx2_evb.yaml
index 24461070291..45eb44557a4 100644
--- a/boards/riscv/it82xx2_evb/it82xx2_evb.yaml
+++ b/boards/riscv/it82xx2_evb/it82xx2_evb.yaml
@@ -13,6 +13,7 @@ supported:
   - adc
   - espi
   - gpio
+  - usb_device
   - i2c
   - kscan
   - pm
```

itebj commented on boards/riscv/it82xx2_evb/it82xx2_evb.dts:
```diff
@@ -197,6 +197,13 @@
 	pinctrl-names = "default";
 };
 
+&usb0 {
```
Got it, sir. Thanks for reminding and guidance. I add the zephyr_udc0 nodelabel back in this upload, and the usb_device has been added as well. Thanks again.

itebj commented:
Dear sirs,

The latest update add the `zephyr_udc0` back and also `usb_device` in the boards/riscv/it82xx2_evb/it82xx2_evb.yaml. 
Thanks again for your revisiting.

itebj commented:
Dear sirs,

In the latest update, we add the `zephyr_udc0` back and we also add `usb_device` in the boards/riscv/it82xx2_evb/it82xx2_evb.yaml. 
Thanks again for your revisiting.

carlescufi commented:
@itebj please rebase to fix conflicts

Commit:
ITE: drivers/usb/device: Add USB Device Controller Support

Add USB Device Driver (usb_dc) of ITE IT82xx2

TEST=west build -p always -b it82xx2_evb
1. zephyr/sample/subsys/usb/hid
2. zephyr/sample/subsys/usb/hid-mouse

Signed-off-by: BJ Chen <bj.chen@ite.com.tw>

itebj commented:
> @itebj please rebase to fix conflicts

Dear sirs,

I just couldn't upload anymore when I try to git push, the error messages was shown as follows:
```
>>>  git push zephyr-fork z-usb -f
Username for 'https://github.com': bj.chen@ite.com.tw
Password for 'https://bj.chen@ite.com.tw@github.com': 
Enumerating objects: 59317, done.
Counting objects: 100% (59314/59314), done.
Delta compression using up to 4 threads
Compressing objects: 100% (21709/21709), done.
Writing objects: 100% (52717/52717), 16.30 MiB | 18.38 MiB/s, done.
Total 52717 (delta 39103), reused 41279 (delta 29057), pack-reused 0
error: RPC failed; HTTP 408 curl 22 The requested URL returned error: 408
send-pack: unexpected disconnect while reading sideband packet
fatal: the remote end hung up unexpectedly
Everything up-to-date
```

I am not pretty sure what just happened but maybe it's about my clicking the "Resolving Conflicts" button.

itebj commented:
Dear sirs,

Well, it seems that the git pushing work again, thanks again for your revisiting.

Best regards,
BJ
